crates/ruff_linter/src/rules/flake8_pyi/rules/generic_not_last_base_class.rs

```diff
@@ -0,0 +1,174 @@
+/// In a stub file, however, this rule is enforced purely for stylistic
+/// consistency.
+///
+/// For example:
+/// ```python
+/// class LinkedList(Generic[T], Sized):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Generic[K, V],
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+/// ):
+///     ...
+/// ```
+///
+/// ```python
+/// class LinkedList(Sized, Generic[T]):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+///     Generic[K, V],
+/// ):
+///     ...
+/// ```
+/// ## References
+/// - [`typing.Generic` documentation](https://docs.python.org/3/library/typing.html#typing.Generic)
+///
+/// [1]: https://github.com/python/cpython/issues/106102
+#[violation]
+pub struct GenericNotLastBaseClass;
+
+impl Violation for GenericNotLastBaseClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`Generic[]` should always be the last base class")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Move `Generic[]` to be the last base class".to_string())
+    }
+}
+
+/// PYI059
+pub(crate) fn generic_not_last_base_class(
+    checker: &mut Checker,
+    class_def: &StmtClassDef,
+    bases: Option<&Arguments>,
+) {
+    let Some(bases) = bases else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+
+    let generic_base_indices: Vec<usize> = bases
+        .args
+        .iter()
+        .enumerate()
+        .filter_map(|(base_index, base)| {
+            if is_generic(base, semantic) {
+                return Some(base_index);
+            }
+            None
+        })
+        .collect();
+
+    if generic_base_indices.is_empty() {
+        return;
+    }
+
+    if generic_base_indices.len() == 1 {
+        let base_index = generic_base_indices[0];
+        if base_index == bases.args.len() - 1 {
+            // Don't raise issue for the last base.
+            return;
+        }
+
+        let mut diagnostic = Diagnostic::new(GenericNotLastBaseClass, class_def.identifier());
+        diagnostic.set_fix(generate_fix(bases, base_index, checker.locator()));
+        checker.diagnostics.push(diagnostic);
+    } else {
+        // No fix if multiple generics are seen in the class bases.
+        checker.diagnostics.push(Diagnostic::new(
+            GenericNotLastBaseClass,
+            class_def.identifier(),
+        ));
+    }
+}
+
+/// Return `true` if the given expression resolves to `typing.Generic[...]`.
+fn is_generic(expr: &Expr, semantic: &SemanticModel) -> bool {
+    if !semantic.seen_typing() {
+        return false;
+    }
+
+    let Expr::Subscript(ast::ExprSubscript { value, .. }) = expr else {
+        return false;
+    };
```

Comment:
We have a helper function for this kind of thing, and I was about to suggest that you use it:

https://github.com/astral-sh/ruff/blob/349a4cf8cea9cfa5bfdd78e94c40220886008b7a/crates/ruff_python_ast/src/helpers.rs#L662-L673

_But_ if you did use that helper function, it would do a slightly different thing here to the PR's current behaviour: it would _also_ flag places where bare `typing.Generic` (unsubscripted with any type parameters) was used as a base class. That always causes an exception at runtime... but maybe it should be flagged by this rule anyway? I can't think of a good reason to _exclude_ bare `typing.Generic` from this check, even if it's going to cause other problems as well. Anyway, I don't feel too strongly on this point, but whichever way you decide to go -- it would be great to add a test that shows the current behaviour if bare `typing.Generic` appears in the bases tuple.

Reply:
I was thinking maybe one can subclass `Generic` by itself to do some metaprogramming type stuff, but since that's straight up `TypeError`, it'll probably be okay to go with your suggestion. Thanks for pointint it out!

Reply:
inheriting from plain `Generic` is specifically allowed if the name of the subclass is "Protocol" FYI:

```pycon
>>> from typing import Generic
>>> class Protocol(Generic): pass
... 
>>> 
```

but I didn't tell you that